[OldDot Rules Migration] Tag rules

BrtqKr opened the pull request:
**Holding on:**

- [x] https://github.com/Expensify/App/pull/47468 (to avoid conflicts)
- [x] https://github.com/Expensify/Auth/pull/12370 (deploy)

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/47016
$ https://github.com/Expensify/App/issues/48977
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to Profile => workspaces => select a workspace => tags
2. Select an existing tag, or create a new one
3. Change approver
4. Verify that selecting the same person, removes them as an approver
5. Verify that there are correct system messages being sent to the admin channel

Verify that disabling rules in more features, disables approver as well 

<img width="493" alt="image" src="https://github.com/user-attachments/assets/0625f0e9-696b-4e85-b1ab-8ea5cc848951">

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/e9b86e26-76a5-44e7-8668-d689575a4aaa

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/29f784e9-fa5d-4070-8ca9-4cbba4748b92

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/d29d7cfd-a258-4b65-bd46-54b25e4500be

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/13df2040-51d2-4ef7-8e78-4c321c803c93

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/19b6beff-a30d-4128-b531-786ef5df539f

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/6792e782-f9e2-4734-abd4-81dacf35ff56

</details>

WojtekBoman committed:
Add WorkspaceMembersSelectionList

BrtqKr committed:
add interface for tag edit

BrtqKr committed:
wire up the tags

BrtqKr committed:
wire up action

marcaaron commented:
The PRs we were waiting on are now deployed or merged so we can take this off HOLD.

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

BrtqKr committed:
cleanup, modify action

BrtqKr committed:
change translations

BrtqKr commented:
@marcaaron, it seems to be in a different field than I initially expected, but I've applied the changes and it's working. I believe we can pass it on for review.

rojiphil commented:
@BrtqKr I will review the PR today. Meanwhile, can you please resolve the conflicts? Thanks

rojiphil reviewed (commented):
@BrtqKr I have a few observations as mentioned below but the overall code looks super good though:
1. Once an approver is selected, we do not allow deselecting the `approver` to set it to `none`. I think we need to support this considering that OldDot supports this.
2. The test steps in PR mentions this:
> Verify that disabling tags in more features, disables approver as well 

Neither do I see any implementation for this nor do I see any reference to this in test videos. Have we missed implementing this? Also, what would `disables approver` mean here? I think it will help QA if we can further break down the test steps

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

BrtqKr committed:
add remove approver option

BrtqKr commented:
@rojiphil 
1. I've added the removal option
2. That's an error in the description, I meant rules and that's already implemented 


```
                    {policy?.areRulesEnabled && (
                        <>
                            <View style={[styles.mh5, styles.mv3, styles.pt3, styles.borderTop]}>
                                <Text style={[styles.textNormal, styles.textStrong, styles.mv3]}>{translate('workspace.tags.tagRules')}</Text>
                            </View>
                            <OfflineWithFeedback pendingAction={policy.pendingFields?.rules}>
                                <MenuItemWithTopDescription
                                    title={tagApprover ?? ''}
                                    description={translate(`workspace.tags.approverDescription`)}
                                    onPress={navigateToEditTagApprover}
                                    shouldShowRightIcon
                                />
                            </OfflineWithFeedback>
                        </>
                    )}
```

rojiphil reviewed (commented):
> 2\. That's an error in the description, I meant rules and that's already implemented

@BrtqKr Thanks for the changes. The removal feature works fine but the feature of disabling approvers does not seem to be implemented yet as we do for category approver rules. Here is a test video demonstrating this. Please have a look. Thanks


https://github.com/user-attachments/assets/15dad6a9-cb39-4958-8172-d39ffb3b61ec

BrtqKr committed:
disable when workflows are disabled

BrtqKr committed:
add approver conditon

BrtqKr committed:
add approver conditon to category settings

BrtqKr commented:
if I understand correctly the condition should probably be uniform and it's been missing for the categories. It's already applied

rojiphil commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>MacOS: Chrome / Safari</summary>

TODO

</details>

<details>
<summary>MacOS: Desktop</summary>

TODO

</details>

<details>
<summary>iOS: mWeb Safari</summary>

TODO

</details>

<details>
<summary>Android: Native</summary>

TODO

</details>

<details>
<summary>Android: mWeb Chrome</summary>

TODO

</details>

<details>
<summary>iOS: Native</summary>

TODO

</details>

rojiphil commented:
> if I understand correctly the condition should probably be uniform and it's been missing for the categories. It's already applied with the remaining fixes

Well! Not actually. What I meant was that the `approval` disabling feature was missing in `tag rules` and that is added now.  Thanks for including that.

But I am not sure if enabling `Workflows—>Add Approvals` is mandatory for `Category/Tag` approver rules. I am not able to interpret this in the [design document](https://docs.google.com/document/d/1oLr14YhL6Y0N5g4tbozdIIrFbybBlsRA0H9I8Wm--w8/edit#bookmark=id.noull64ot4cc). And it seems strange that we should force the user to enable `Add approvals` in workflows just for the sake of `category/tag` approval rules even though `approvals` were not required for payments. However, the `hint` text indicates that enabling `Workflow—>Add Approvals` is mandatory. 

@marcaaron What’s your take on this?

![Screenshot 2024-09-11 at 12 18 53 PM](https://github.com/user-attachments/assets/702ad8f6-f32b-41ac-b7f3-dab9b71b8ab7)

![Screenshot 2024-09-11 at 12 19 14 PM](https://github.com/user-attachments/assets/3892023f-4de0-47ae-a570-ce9fc1882944)

rojiphil commented:
@BrtqKr There is an issue I notice with `Category` approval rules. Since we only check for `value` [here](https://github.com/Expensify/App/blob/762bffcfad4ca7c4553eabedde5a2ba47871bdb5/src/libs/CategoryUtils.ts#L48) and [here](https://github.com/Expensify/App/blob/762bffcfad4ca7c4553eabedde5a2ba47871bdb5/src/libs/actions/Policy/Category.ts#L1153), we could end up with wrong values as demonstrated below. We need to additionally check for `field` in categories like we do for tags to avoid any issues. Same changes can be applied [here](https://github.com/Expensify/App/blob/762bffcfad4ca7c4553eabedde5a2ba47871bdb5/src/libs/CategoryUtils.ts#L52) and [here](https://github.com/Expensify/App/blob/762bffcfad4ca7c4553eabedde5a2ba47871bdb5/src/libs/actions/Policy/Category.ts#L1232) too. Although not directly related to `tag` rules, we may want to fix this here as it is related to the overall OldDot rules migration project.

![Screenshot 2024-09-11 at 1 30 23 PM](https://github.com/user-attachments/assets/423d123a-4104-4d34-a26c-0e727bd681c4)

rojiphil reviewed (commented):
@BrtqKr Although NAB, I have left a comment. Please have a look. Otherwise, the code LGTM.
Once the comments [here](https://github.com/Expensify/App/pull/48325#issuecomment-2342810724) and [here](https://github.com/Expensify/App/pull/48325#issuecomment-2343061930) are also addressed, I will complete the checklist

BrtqKr committed:
replace values with const

BrtqKr committed:
change conditon for find category approver

BrtqKr committed:
ts fix

BrtqKr commented:
> > if I understand correctly the condition should probably be uniform and it's been missing for the categories. It's already applied with the remaining fixes
> 
> Well! Not actually. What I meant was that the `approval` disabling feature was missing in `tag rules` and that is added now. Thanks for including that.
> 
> But I am not sure if enabling `Workflows—>Add Approvals` is mandatory for `Category/Tag` approver rules. I am not able to interpret this in the [design document](https://docs.google.com/document/d/1oLr14YhL6Y0N5g4tbozdIIrFbybBlsRA0H9I8Wm--w8/edit#bookmark=id.noull64ot4cc). And it seems strange that we should force the user to enable `Add approvals` in workflows just for the sake of `category/tag` approval rules even though `approvals` were not required for payments. However, the `hint` text indicates that enabling `Workflow—>Add Approvals` is mandatory.
> 
> @marcaaron What’s your take on this?
> 
> ![Screenshot 2024-09-11 at 12 18 53 PM](https://private-user-images.githubusercontent.com/3004659/366332315-702ad8f6-f32b-41ac-b7f3-dab9b71b8ab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNTQ1MjcsIm5iZiI6MTcyNjA1NDIyNywicGF0aCI6Ii8zMDA0NjU5LzM2NjMzMjMxNS03MDJhZDhmNi1mMzJiLTQxYWMtYjdmMy1kYWI5YjcxYjhhYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMTEzMDI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDUwMGFlNWNjY2QzYTEzYWQ1ZjIwYjYxYTlkM2NlMjdmMTZmNjEwZGIzMDBjZDc2Mjk5ZTUyMmQwOTU2NDA2ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.j_SEObT3fEoinkRgqL14_cJHfrasg-gBI6QXHodBKxs)
> 
> ![Screenshot 2024-09-11 at 12 19 14 PM](https://private-user-images.githubusercontent.com/3004659/366332365-3892023f-4de0-47ae-a570-ce9fc1882944.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNTQ1MjcsIm5iZiI6MTcyNjA1NDIyNywicGF0aCI6Ii8zMDA0NjU5LzM2NjMzMjM2NS0zODkyMDIzZi00ZGUwLTQ3YWUtYTU3MC1jZTlmYzE4ODI5NDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMTEzMDI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODg4OTlhYzRiMDk0NWJiYzVlMTY1NTA4Njk2ZDUxYjlkYTQ0Njk1ZmJmNmNmZmVmYjY3MGE3MjcxYjI5YThlZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.hHVwKP1Bxjn81KBkc2vBSaAWUwpSEuXM3RhrBP4n0Yc)

@marcaaron this is probably the only matter left - is it necessary and should be uniform for both approval types, or should this be only category-specific and we should only disable tag approver section based on whether the workflows are enabled, or not

marcaaron commented:
>I am not sure if enabling Workflows—>Add Approvals is mandatory for Category/Tag approver rules. I am not able to interpret this in the [design document](https://docs.google.com/document/d/1oLr14YhL6Y0N5g4tbozdIIrFbybBlsRA0H9I8Wm--w8/edit#bookmark=id.noull64ot4cc). 

For category "Approver" rule, yes. That's stated already in the Design Doc. Though, it's a good questions about whether the same is true for Tags. I'll defer to @JmillsExpensify on that. We set it up this way since most customer who use these features also have advanced approval settings. Only 1% of users use these features so my gut says we can optimize this later if necessary.

rojiphil commented:
> In other words, let's please not block on this one

Sure. Let me finish the checklist now then.

BrtqKr committed:
use getCategoryApproverRule in the actions

BrtqKr committed:
change offline pattern for tags

BrtqKr commented:
@rojiphil, We've discussed offline pattern for rules with @WojtekBoman. It's already applied for both categories and tags and we'd like to keep it uniform. It seems to be working as expected.

https://github.com/user-attachments/assets/d9646cba-c354-415b-bcf5-a856458e90e8

BrtqKr committed:
ts fix

rojiphil commented on src/libs/actions/Policy/Category.ts:
```diff
@@ -1194,9 +1196,9 @@ function setPolicyCategoryApprover(policyID: string, categoryName: string, appro
-                            approvalRules: null,
+                    pendingRulesUpdates: {
+                        [categoryName]: {
+                            approvalRule: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
```
Let's reset `approvalRule` to `null` here

rojiphil commented on src/libs/actions/Policy/Category.ts:
```diff
@@ -1209,8 +1211,10 @@ function setPolicyCategoryApprover(policyID: string, categoryName: string, appro
+                    },
+                    pendingRulesUpdates: {
+                        [categoryName]: {
+                            approvalRule: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
```
Same as above. Let's reset `approvalRule` to `null` here

rojiphil commented on src/libs/actions/Policy/Category.ts:
```diff
@@ -1182,8 +1182,10 @@ function setPolicyCategoryApprover(policyID: string, categoryName: string, appro
-                            approvalRules: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
+                    },
+                    pendingRulesUpdates: {
+                        [categoryName]: {
```
Let's include `field` name too along with the `categoryName` to ensure uniqueness. Otherwise, this can clash with `tag` names

BrtqKr replied:
nice catch, I've added a utility function for that

rojiphil reviewed (commented):
@BrtqKr I have left few comments. Please have a look. Looks like we are pretty close to getting this done.

BrtqKr committed:
make pending keys unique

marcaaron reviewed (changes requested):
Looks great so far had a few small comments.

rojiphil commented:
Oh! Just noticed that our current logic will reset the tag rules `approver` when we edit the `tag` name. This is because we use `tag` name to identify the `approver`. The same concern applies to `category` too. 
This means we must update the rules with the modified name whenever we edit the `tag` name. 
But I also wonder if it is better to avoid depending on the `tag` name to identify the `approver`.

BrtqKr commented on src/libs/actions/Policy/Tag.ts:
```diff
@@ -846,6 +847,90 @@ function setPolicyTagGLCode(policyID: string, tagName: string, tagListIndex: num
+    const approverRuleToUpdate = PolicyUtils.getTagApproverRule(policyID, tag);
+    const filteredApprovalRules = approverRuleToUpdate ? prevApprovalRules.filter((rule) => rule.id !== approverRuleToUpdate.id) : prevApprovalRules;
+    const toBeUnselected = approverRuleToUpdate?.approver === approver;
+    const tagRuleUpdateKey = PolicyUtils.getPolicyFieldUpdateKey(tag, CONST.POLICY.FIELDS.TAG);
```
I thought about this, but if we're trying to use it for multiple use cases (categories, tags etc) then this might enforce this to be uniform in all of them and avoid a typo when copypasting it inside of actions, offline wrappers and so on. If it wasn't for the multiple file, where those keys are being used, I'd avoid it. I don't have a strong opinion on that, so up to you 🤷

marcaaron replied:
I'm good with it as long as it has a more descriptive name for the function and some explanation of how to use it.

BrtqKr commented:
@rojiphil, unless we get a specific ID per tag/category, we cannot identify it properly in this particular scenario. We need to rely on something higher in the hierarchy, otherwise we might end up with id conflicts.

> This means we must update the rules with the modified name whenever we edit the tag name.

Regarding this thing, this is quite unintuitive to have to update this manually in so many places. I'd expect potential errors during the maintenance quite soon. I'd suggest asking back-end team for the model changes to include proper ids for the tag/category, but I don't think we should be holding this ticket until then.

BrtqKr committed:
review fixes

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

marcaaron commented:
>our current logic will reset the tag rules approver when we edit the tag name. This is because we use tag name to identify the approver. The same concern applies to category too.

>Regarding this thing, this is quite unintuitive to have to update this manually in so many places. I'd expect potential errors during the maintenance quite soon. 

For now, we could also do some solution where if it's "optimistic" i.e. there is no `id` yet we should just not let people edit the tag or category name.

>I'd suggest asking back-end team for the model changes to include proper ids for the tag/category, but I don't think we should be holding this ticket until then.

I would worry that after a certain point we are engineering features that people might not actually need or want to do offline. It would take careful consideration e.g. the "id" would need to be generated by the client. And I don't think it's worth changing at this stage since the overhead is quite significant.

rojiphil commented:
> > our current logic will reset the tag rules approver when we edit the tag name. This is because we use tag name to identify the approver. The same concern applies to category too.
> 
> > Regarding this thing, this is quite unintuitive to have to update this manually in so many places. I'd expect potential errors during the maintenance quite soon.
> 
> For now, we could also do some solution where if it's "optimistic" i.e. there is no `id` yet we should just not let people edit the tag or category name.

Yeah. That makes sense as an immediate solution to not allow edit of tag or category name if `id` is invalid (i.e. when offline). 
When online though, the `approver` would be reset momentarily until the BE sends back the updated data but I think that’s fine as it would be an overkill to address this at the moment.

@BrtqKr The branch has got conflicts. Please resolve this too. Thanks.

marcaaron commented:
@BrtqKr Let us know when this is ready for another review. Seems @rojiphil has some of the checklist left to do, but I think they are waiting for the latest comments to get resolved. Let me know if there's anything I can do to help unblock!

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

BrtqKr commented:
@marcaaron @rojiphil 
I'm not sure if this suggestion makes sense. If I understand correctly: 
1. The main issue is that after renaming the category its content is moved to the new entry in the dictionary
2. Since the rules are operating on the category name as a key, there is no relation between them, so the approver is disappearing 

But...adding any id-based condition for disable (I'm not sure what kind of id you mean here, because neither category nor tag has one) without having them saved with a proper relation is not going to work, because at some point we're not going to be able to differentiate what the rule was supposed to apply to. 

It's not a matter of the optimistic update, even a single change to the category name or tag name breaks the relation between the rules and categories. The only things I can do at this point is to:
1. Allow to edit a category once(because we have a previous name)and then completely disable it
2. Completely disable rename in the offline mode
3. Try to somehow relate everything in the actions, but this is going to be really messy and would cause bugs.

marcaaron commented:
Here’s what I am going to propose for now:

- When category or tag name changes call `getCategoryApproverRule()` or `getTagApproverRule()`
- Update the rule with the new name. They are linked by the name so I assume this should work even when we are offline and should fix the issue of the "approver" not transferring.
- Let’s punt the conversation about “pending” stuff for now and also scrap the current changes - essentially not providing feedback for the pending stuff. I think it did not get enough consideration in the design phase. I also think that we need to re-think the design. Looking at what we have currently and I’m not sure I like it. But, it's ok, not having the feedback is also not a huge deal for now. So, let's just remove those parts and keep what we like?

At some point later we can explore:

- Better offline handling.
- Different data structure (perhaps we can consolidate the related data so it is easier to update and set errors and also switch to objects instead of arrays for "rules" since they are harder to work with).

BrtqKr committed:
update category rule when renaming category

BrtqKr committed:
update tag rule when renaming tag

BrtqKr committed:
remove pending action pattern for rules

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

BrtqKr committed:
fix onyx check

BrtqKr committed:
remove pending pattern from type

BrtqKr commented:
I've removed the offline pattern and added category/tag renaming in the rules. I believe this would be everything

https://github.com/user-attachments/assets/7366be3c-efce-47f9-9770-2a472f214be0

BrtqKr committed:
run prettier

BrtqKr committed:
cleanup

BrtqKr committed:
cleanup

BrtqKr committed:
remove withOnyx

rojiphil commented on src/libs/actions/Policy/Policy.ts:
```diff
@@ -2578,7 +2578,10 @@ function createWorkspaceFromIOUPayment(iouReport: OnyxEntry<Report>): WorkspaceF
 
     // We need to move the report preview action from the DM to the workspace chat.
-    const reportPreview = ReportActionsUtils.getParentReportAction(iouReport);
+    const parentReport = ReportAcionsConnection.getAllReportActions()?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${iouReport.parentReportID}`];
```
@BrtqKr Not sure if I am missing something here but why did we have to create `ReportAcionsConnection` when we already have a utility `ReportActionsUtils.getAllReportActions`?

BrtqKr replied:
because we have a new eslint rule, which is forcing use to fix those things when we edit files containing deprecated code

rojiphil replied:
@BrtqKr Ah! I see. So, the deprecated function is `getParentReportAction` but is there a need to create `ReportAcionsConnection` when `ReportActionsUtils.getAllReportActions` gives us the same result?

BrtqKr replied:
This is something I'll need to discuss separately, but from what I understand the main issue is that the rule has been prepared with components in mind, so it's expecting some kind of action to be taken anyway. I was thinking about the possible benefits of having a connection in one place instead of having logic split between different files, but as I said earlier, I'll discuss it in more detail with someone who has more context.

marcaaron replied:
This doesn't feel like a blocker.

rojiphil replied:
> This is something I'll need to discuss separately

`getParentReportAction` has been used in many places and I agree that this needs a separate discussion. Since this is not at all related to this issue, I would prefer not to include the changes for this in this PR. As and when there is a conclusion on the way forward, we can add the code changes. Anyway, I would leave the final decision on this to @marcaaron. 

If we are leaving the code as it is, let us correct the spelling `ReportAcionsConnection`.

rojiphil commented on package-lock.json:
```diff
@@ -27979,22 +27979,6 @@
         "jest": "bin/jest.js"
       }
     },
-    "node_modules/jest-expo/node_modules/@babel/code-frame": {
```
@BrtqKr Did this change got introduced unintended as these seem unrelated to our current scope here?

BrtqKr replied:
I haven't modified anything in package.json and I'm using the appropriate node version. The change was introduced right after merging main to this branch and reinstalling the packages, so I'd assume it wasn't indended, but it's correct.

marcaaron replied:
This also doesn't seem like a blocker.

rojiphil commented:
Otherwise the code looks good and the feature works well. 
I have left a couple of comments. Once we address them, I can do the final testing.
Looks very close to getting this done.

BrtqKr committed:
Merge remote-tracking branch 'origin/main' into brtqkr/47016-tag-rules

BrtqKr committed:
run prettier

BrtqKr committed:
cleanup

rojiphil reviewed (approved):
@BrtqKr I have left few suggestions related to test cases 

1. Let us add the below mentioned steps from 5th step onward in the existing test steps.
- Select an approver again 
- Change the name of the tag
- Verify that the `approver` does not change.

2. Also, let us add the following so that QA can test `categories` too

`Repeat the above mentioned steps for categories and verify.`

3. Let us remove the following text from steps as QA will get confused:

`Verify that disabling rules in more features, disables approver as well`

@marcaaron Other than [this comment](https://github.com/Expensify/App/pull/48325#discussion_r1766240465), the code changes related to tag/category approver feature LGTM and works well too and I have completed the checklist.
Over to you for review.

BrtqKr committed:
rename connection

marcaaron commented on src/languages/es.ts:
```diff
@@ -3233,6 +3233,8 @@ export default {
             glCode: 'Código de Libro Mayor',
             updateGLCodeFailureMessage: 'Se produjo un error al actualizar el código de Libro Mayor. Por favor, inténtelo nuevamente.',
+            tagRules: 'Reglas de etiquetas',
+            approverDescription: 'Aprobador',
```
NAB we could probably move this to the `general` since it is used twice.

marcaaron reviewed (approved):
LGTM

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/10950314341) to staging by https://github.com/marcaaron in version: 9.0.39-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/11020071968) to production by https://github.com/AndrewGable in version: 9.0.39-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

QichenZhu commented on src/pages/workspace/tags/TagApproverPage.tsx:
```diff
@@ -0,0 +1,57 @@
+    const styles = useThemeStyles();
+    const {translate} = useLocalize();
+
+    const tagApprover = PolicyUtils.getTagApproverRule(policyID, tagName)?.approver;
```
This page didn’t re-render when the policy updates. In issue https://github.com/Expensify/App/issues/57514, we want to see the checkmark immediately after selecting an approver, even though the page closes soon after. Fixed by using the `usePolicy()` hook.
